src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentChecker.java

```diff
@@ -0,0 +1,125 @@
+                            // Get the value of the field and process the list
+                            processList((List<?>) field.get(hcd), allDocuments);
+                        }
+                    } else if (DocumentFileNameProvider.class.isAssignableFrom(field.getType())) {
+                        allDocuments.add((DocumentFileNameProvider) field.get(hcd));
+                    } else if (HasCaseDocument.class.isAssignableFrom(field.getType())) {
+                        processHasCaseDocument((HasCaseDocument)  field.get(hcd), allDocuments);
+                    }
+                }
+            } catch (Exception e) {
+                log.error("Failed to check for duplicate filenames and return warnings", e);
+            }
+        }
+    }
+
+    private List<DocumentFileNameProvider> collectCaseDocumentsFromFinremCaseData(FinremCaseData caseData) {
+        List<DocumentFileNameProvider> allDocuments = new ArrayList<>();
+        processHasCaseDocument(caseData, allDocuments);
+        return allDocuments.stream().filter(Objects::nonNull).toList();
+    }
+
+    @Override
+    public List<String> getWarnings(CaseDocument caseDocument, byte[] bytes, FinremCaseDetails beforeCaseDetails, FinremCaseDetails caseDetails)
+        throws DocumentContentCheckerException {
+        FinremCaseData caseData = beforeCaseDetails.getData();
+        List<DocumentFileNameProvider> allDocuments = collectCaseDocumentsFromFinremCaseData(caseData);
+
+        log.info("Iterating all CaseDocuments with interface HasCaseDocument.");
+
+        // Check for duplicate filenames in the collected documents
+        boolean hasDuplicates = allDocuments.stream()
+            .anyMatch(d -> isDuplicateFilename(caseDocument, () -> List.of(d)));
+
+        if (hasDuplicates) {
+            return List.of(WARNING);
+        }
+
+        // check if duplication occurs on uploading documents
+        FinremCaseData newCaseData = caseDetails.getData();
+        if (
+            (newCaseData.isConsentedApplication()
+                && ofNullable(newCaseData.getUploadDocuments()).orElse(List.of()).stream()
+                    .map(d -> ofNullable(d.getValue()).orElse(UploadDocument.builder().build()).getDocumentLink())
+                    .filter(Objects::nonNull)
+                    .filter(d -> isDuplicateFilename(caseDocument, () -> List.of(d))).count() > 1)
+                ||
+                (newCaseData.isContestedApplication()
+                    && ofNullable(newCaseData.getUploadGeneralDocuments()).orElse(List.of()).stream()
+                        .map(d -> ofNullable(d.getValue()).orElse(UploadGeneralDocument.builder().build()).getDocumentLink())
+                        .filter(Objects::nonNull)
+                    .filter(d -> isDuplicateFilename(caseDocument, () -> List.of(d))).count() > 1)
+        ) {
+            return List.of(WARNING);
+        }
+
+        return Collections.emptyList();
+    }
+}
```

Comment:
I notice the getWarnings here checks beforeCaseDetails and newCasedata.  I _think_ the unit tests just test newCaseData?  

The unit tests implicitly test these private functions, so sorry if I missed it.

Reply:
checking on `newCaseData` only happens when I check the duplication in new files which is covered in - 
`testGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeConsented`
`shouldNotGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeConsented`
`testGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeContested`
`shouldNotGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeContested`